### PR TITLE
Reorganize objdb and add basic consul support

### PR DESCRIPTION
**Pull request by shaleman**
@mapuri I ended up reorganizing objdb code to simplify it a bit.
Mainly i merged etcdplugin/consulplugin into single package to avoid recursive import problems.
Now code is simpler and API is simpler too.
PTAL.

Regd golang benchmarks, I added some benchmark tests. But, golang benchmarks doesn't allow me to benchmark reads and write in isolation. Basically, i want to benchmark writing 1K unique keys, reading 1K unique keys and deleting 1K unique keys. The way golang benchmark works, i have to write unique keys before i can read it. so, i can't benchmark reads in isolation.

### CODE REVIEW

File: client/client_test.go [hunk 1]
```diff
@@ -113,6 +113,108 @@ func TestSetGetPerformance(t *testing.T) {
...
+		}
+	}
+
+	timeTook := time.Since(startTime).Nanoseconds() / 1000000
+	log.Infof("Write Test took %d milli seconds per write. %d ms total", timeTook/testCount, timeTook)
+
+	log.Infof("Performing %d read tests", testCount)
+
+	// Get test
+	startTime = time.Now()
```
Comment (mapuri) [inline client/client_test.go]: I think we can use golang benchmarks (https://golang.org/pkg/testing/#B) instead to measure performance.

**Comment by mapuri** [comment]
LGTM, other than one minor comment. 

Also just minor, I see that you we have tests written but didn't find a make target for those. Are theses manually run? I was wondering if we should setup a CI job for this repo. I can help.

**Comment by shaleman** [comment]
@mapuri Yes, There is no direct CI for these. 
We have few more such repos which require etcd or ovs for unit tests but don't have good CI coverage.
Will it be possible to create a catchall CI job that pulls these repos in a vagrant environment and run their unit tests..

**Comment by mapuri** [comment]
@shaleman 

> We have few more such repos which require etcd or ovs for unit tests but don't have good CI coverage.

what are those repos? Do they have a `make test` target that we can use for CI? If we have test jobs for each repo (need not be vagrant, it can be simple `go test` target sometimes), we can setup independent CI jobs.

> Will it be possible to create a catchall CI job that pulls these repos in a vagrant environment and run their unit tests..

It might be good to hook CI to a PR, as it helps validating changes to the repo in just that PR. Doing it together for multiple repos might not be ideal as we might end up pulling multiple sets of untested changes.
 
If we really need a common CI, we can perhaps put together a job (like a nightly sanity) that can pull all these repos and run a test. But I am not sure if it will be much useful as it won't catch issues close to the PR and might actually be burdensome to debug failures.

**Comment by shaleman** [comment]
Added Vagrantfile, makefile, Godeps and other plumbing so that i can setup a Jenkins job for the repo

**Comment by mapuri** [comment]
couldn't review Vagrantfile and Makefile changes due to them being in same commit as godeps. Otherwise LGTM.

**Comment by shaleman** [comment]
@mapuri 
Thanks. I didn't know about b.ResetTimer().
I had this comment about not using go benchmarks:

Regd golang benchmarks, I added some benchmark tests. But, golang benchmarks doesn't allow me to benchmark reads and write in isolation. Basically, i want to benchmark writing 1K unique keys, reading 1K unique keys and deleting 1K unique keys. The way golang benchmark works, i have to write unique keys before i can read it. so, i can't benchmark reads in isolation.

But, May be I can workaround it by using b.ResetTimer(). Let me check that.

btw, Here is the diff excluding the Godeps: https://gist.github.com/shaleman/bbf683c61839ebc75ba3

**Comment by mapuri** [comment]
cool sgtm @shaleman .

Regarding Vagrantfile for test environment, you might want to consider using Dockerfile and test image instead for running tests. It will be much lighter weight but up to you.

**Comment by shaleman** [comment]
Thanks @mapuri 
b.ResetTimer() worked great. I moved the tests to go benchmark.

I wasn't sure how to run etcd/consul inside the container and then run the go tests. Using vagrant was a easy way out :)

**Comment by mapuri** [comment]
@shaleman cool, thanks for trying out the benchmark package 

LGTM

**Pull request merged**